src/net/connection/quality.rs

```diff
@@ -1,18 +1,93 @@
-// TODO: Congestion avoidance.
-// To prevent congestion avoidance we need ot decide whether the connection is good or bad.
-// If the network of the client is bad we do not flood the router with small packets.
-// When network conditions are `Good` we send 30 packets per-second, and when network conditions are `Bad` we drop to 10 packets per-second.
+use std::time::Duration;
+use std::sync::RwLockWriteGuard;
+
+use net::connection::VirtualConnection;
+
+use net::NetworkConfig;
+use packet::CongestionData;
 
 /// Represents the quality of an network.
 pub enum NetworkQuality {
     Good,
     Bad,
 }
 
+/// This type helps with calculating the round trip time from any packet.
+/// It is able to smooth out the network jitter if there is any.
+pub struct NetworkQualityMeasurer {
+    config: NetworkConfig,
+}
+
+impl NetworkQualityMeasurer
+{
+    pub fn new(config: NetworkConfig) -> NetworkQualityMeasurer
+    {
+        NetworkQualityMeasurer { config }
+    }
+
+    /// This will calculate the round trip time (rtt) from the given acknowledgement.
+    /// Where after it updates the rtt from the given connection.
+    pub fn update_connection_rtt(&self, connection: &mut RwLockWriteGuard<VirtualConnection>, ack_seq: u16)
```

Comment:
Huh, this is interesting. Just out of curiosity, why not pass in an arc?

Reply:
This was because this function gets called when the lock is already requested on the connection. So I thought why not directly pass in an unlocked connection into this function.

Reply:
Yeah, that's pretty cool. I've never thought to do it that way. Nice.